html/semantics/scripting-1/the-script-element/moving-between-documents.html

```diff
@@ -0,0 +1,132 @@
+        resolve([iframe.contentWindow, iframe.contentDocument]);
+      } else if (documentType === "createHTMLDocument") {
+        resolve([
+            iframe.contentWindow,
+            iframe.contentDocument.implementation.createHTMLDocument("")]);
+      } else {
+        reject("Invalid document type: " + documentType);
+      }
+    };
+    document.body.appendChild(iframe);
+  });
+}
+
+window.scriptErrorEventFired = false;
+window.didExecute = undefined;
+
+(async function() {
+  for (const during of ["parsing", "successful fetch", "failed fetch"]) {
+    for (const destType of ["iframe", "createHTMLDocument"]) {
+      for (const type of ["text/javascript", "module"]) {
+        const description =
+            `Move during ${during} to ${destType} document: ${type}`;
+        const t = async_test(description + ', script evaluation');
+        const tEvent = async_test(description + ', load/error events');
+
+        const [sourceWindow, sourceDocument] = await createDocument("iframe");
+        const [destWindow, destDocument] = await createDocument(destType);
+
+        let scriptErrorEventFired = false;
+
+        sourceWindow.didExecute = undefined;
+        sourceWindow.tEvent = tEvent;
+        destWindow.didExecute = undefined;
+        destWindow.tEvent = tEvent;
+
+        const shouldErrorEventFired =
+          during === "failed fetch" && destType === "iframe";
+
+        if (during === "parsing") {
+          sourceDocument.write("<streaming-element>");
+          destDocument.body.appendChild(
+            sourceDocument.querySelector("streaming-element"));
+          sourceDocument.write(
+            `<script id="s1" type="${type}"
+                     onload="tEvent.unreached_func('onload')"
+                     onerror="tEvent.unreached_func('onerror')">
+            window.didExecute = true;<` + "/script>"
+          );
+          sourceDocument.write("</streaming-element>");
+          sourceDocument.close();
+        } else { // Move during fetching
+          const s = sourceDocument.createElement("script");
+          s.type = type;
+
+          s.onload = tEvent.unreached_func("onload");
+          if (shouldErrorEventFired) {
+            s.onerror = () => { scriptErrorEventFired = true };
+          } else {
+            s.onerror = tEvent.unreached_func("onerror");
+          }
+
+          if (during === "successful fetch") {
+            s.src = "resources/slow-flag-setter.py?result=successful";
+          } else {
+            s.src = "resources/slow-flag-setter.py?result=failed";
+          }
+          sourceDocument.body.appendChild(s);
+          // Need to delay since the "prepare a script" algorithm also
+          // contains related checks; we want to test the
+          // "execute a script block" algorithm for when the fetch comes back.
+          t.step_timeout(() => {
+            destDocument.body.appendChild(s);
+          }, 0);
+        }
+
+        setTimeout(() => {
```

Comment:
Use t.step_timeout here?

Reply:
There are two concurrent tests (`t` and `tEvent`). I preferred very slightly the current code:

```
setTimeout(() => {
  t.step_func_done(...)();
  tEvent.step_func_done(...)();
}, 2000);
```

over:

```
t.step_timeout(() => { ... }, 2000);
tEvent.step_timeout(() => { ... }, 2000);
```

to avoid a slight race condition between two step_timeout()s, but I'm still not so sure.
WDYT?

Reply:
I agree a single timeout is better. Does the global `step_timeout` work? https://web-platform-tests.org/writing-tests/testharness-api.html says

> For single-page tests, step_timeout is also available as a global function.

but I'm not sure if that means it only exists in single-page tests, or if it's primarily used by single-page tests.

Either way it's worth adding a comment explaining why we're using setTimeout and/or global step_timeout.

Reply:
Global `step_timeout()` looks good. Done (and added a comment).